jetty-websocket/websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/FragmentExtension.java

```diff
@@ -69,4 +88,68 @@ public void init(ExtensionConfig config, WebSocketComponents components)
         int maxLength = config.getParameter("maxLength", -1);
         configuration.setMaxFrameSize(maxLength);
     }
+
+    public class FragmentingDemandingFlusher extends DemandingFlusher
+    {
+        public FragmentingDemandingFlusher()
+        {
+            super(FragmentExtension.this::nextIncomingFrame);
+        }
+
+        @Override
+        protected boolean handle(Frame frame, Callback callback, boolean first)
+        {
+            if (first)
+            {
+                if (OpCode.isControlFrame(frame.getOpCode()))
+                {
+                    forwardFrame(frame, callback);
+                    return true;
+                }
+            }
+
+            ByteBuffer payload = frame.getPayload();
+            int remaining = payload.remaining();
+            long maxFrameSize = configuration.getMaxFrameSize();
+            int fragmentSize = (int)Math.min(remaining, maxFrameSize);
+
+            boolean continuation = (frame.getOpCode() == OpCode.CONTINUATION) || !first;
+            Frame fragment = new Frame(continuation ? OpCode.CONTINUATION : frame.getOpCode());
+            boolean finished = (maxFrameSize <= 0 || remaining <= maxFrameSize);
+            fragment.setFin(frame.isFin() && finished);
+
+            if (finished)
+            {
+                // If finished we don't need to fragment, forward original payload.
+                fragment.setPayload(payload);
+            }
+            else
+            {
+                // Slice the fragmented payload from the buffer.
+                int limit = payload.limit();
+                int newLimit = payload.position() + fragmentSize;
+                payload.limit(newLimit);
+                ByteBuffer payloadFragment = payload.slice();
+                payload.limit(limit);
+                fragment.setPayload(payloadFragment);
+                payload.position(newLimit);
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Fragmented {}->{}", frame, fragment);
+            }
+
+            Callback payloadCallback = Callback.from(() ->
+            {
+                if (finished)
+                    callback.succeeded();
+            }, t ->
+            {
+                // This is wrapped with CountingCallback so will only be failed once.
+                callback.failed(t);
+                failFlusher(t);
+            });
```

Comment:
The application doesn't need to succeed this `payloadCallback` until it decides it doesn't need the content anymore. This could be many iterations into this iterating callback where we have a callback for a different frame entirely.